lib/Frontend/Frontend.cpp

```diff
@@ -261,236 +264,305 @@ ModuleDecl *CompilerInstance::getMainModule() {
+    const ImplicitImports &implicitImports) {
+  SharedTimer timer("performSema-addMainFileToModule");
 
-    auto *NextInput = new (*Context) SourceFile(*MainModule,
-                                                SourceFileKind::Library,
-                                                BufferID,
-                                                modImpKind,
-                                                KeepTokens);
-    MainModule->addFile(*NextInput);
-    addAdditionalInitialImports(NextInput);
+  const InputFileKind Kind = Invocation.getInputKind();
+  assert(Kind == InputFileKind::IFK_Swift || Kind == InputFileKind::IFK_SIL);
 
-    if (BufferID == PrimaryBufferID)
-      setPrimarySourceFile(NextInput);
+  if (Kind == InputFileKind::IFK_Swift)
+    SourceMgr.setHashbangBufferID(MainBufferID);
 
-    auto &Diags = NextInput->getASTContext().Diags;
-    auto DidSuppressWarnings = Diags.getSuppressWarnings();
-    auto IsPrimary
-      = PrimaryBufferID == NO_SUCH_BUFFER || BufferID == PrimaryBufferID;
-    Diags.setSuppressWarnings(DidSuppressWarnings || !IsPrimary);
+  auto *MainFile = new (*Context) SourceFile(
+      *MainModule, Invocation.getSourceFileKind(), MainBufferID,
+      implicitImports.kind, Invocation.getLangOptions().KeepTokensInSourceFile);
+  MainModule->addFile(*MainFile);
+  addAdditionalInitialImportsTo(MainFile, implicitImports);
 
-    bool Done;
-    do {
-      // Parser may stop at some erroneous constructions like #else, #endif
-      // or '}' in some cases, continue parsing until we are done
-      parseIntoSourceFile(*NextInput, BufferID, &Done, nullptr,
-                          &PersistentState, DelayedCB.get());
-    } while (!Done);
+  if (MainBufferID == PrimaryBufferID)
+    setPrimarySourceFile(MainFile);
+}
 
-    Diags.setSuppressWarnings(DidSuppressWarnings);
+void CompilerInstance::parseAndCheckTypes(
+    const ImplicitImports &implicitImports) {
+  SharedTimer timer("performSema-parseAndCheckTypes");
+  std::unique_ptr<DelayedParsingCallbacks> DelayedCB{
+      computeDelayedParsingCallback()};
 
-    performNameBinding(*NextInput);
-  }
+  PersistentParserState PersistentState;
 
+  bool hadLoadError = parsePartialModulesAndLibraryFiles(
+      implicitImports, PersistentState, DelayedCB.get());
   if (Invocation.isCodeCompletion()) {
     // When we are doing code completion, make sure to emit at least one
     // diagnostic, so that ASTContext is marked as erroneous.  In this case
     // various parts of the compiler (for example, AST verifier) have less
     // strict assumptions about the AST.
     Diagnostics.diagnose(SourceLoc(), diag::error_doing_code_completion);
   }
-
   if (hadLoadError)
     return;
 
-  // Compute the options we want to use for type checking.
+  OptionSet<TypeCheckingFlags> TypeCheckOptions = computeTypeCheckingOptions();
+
+  // Parse main file last in order to make sure that it can use decls from other
+  // files in the module.
```

Comment:
This is really "type-check main file after parsing all other files so that it can use decls from other files", but with the mess that the main file has parsing and type-checking intertwined.

Reply:
I replaced the comment with something based on your words above.